Feature/control import options

artem-sidorenko opened the pull request:
Import options are displayed by creation of a new project. Currently all possible options are there, which leads to following possible issues:

**Issue 1 - End user confusion: Configuration recommendation outside of own responsibility**
1. Not configured sources(oauth) - which are not usable for the end user of Gitlab installation
2. End user gets probably a hint, which links to the oath configuration steps, which can be done only by Gitlab admin
3. Admin phone is ringing / email inbox gets a new mail / support ticket system gets a new ticket with a support request like "why do we have Github as a source here, but I can't use that?"

**Issue 2: Restricted environments - No internet access:**
1. In some environments with strong security requirements its quite common **NOT** to have Internet access (or not a permanent direct internet connection)
2. In this cases such import functionality won't work at all and can't be realized (because of restrictions)
3. This leads to the Issue 1

**Issue 3: Missing OAuth credentials**
1. Organisation using this particular Gitlab installation doesn't have Github/Bitbucket/\* accounts and has no need to have them
2. For the organisation it doesn't make really sense to create the according accounts only to allow this import
3. This leads to the Issue 1

I've already created the #9553 and here the first code implementation. This MR is marked as [WIP] and I would like to get your feedback in the current state and clarify some questions.
(Please keep in mind, this is my first MR to Gitlab, I want to get feedback and have some ugly code for sure:) )

The idea to solve this issues is to provide a possibility to configure the import sources (and to keep all enabled by default)

Admin interface (new buttons + links to the help pages for admins)
![admin-interface](https://cloud.githubusercontent.com/assets/1645670/9277180/beec227e-42a8-11e5-96bc-f4d572a3d73d.png)

New project interface (same like before, but only the enabled/configured sources are displayed and no admin related help hints for end users)
![new-project](https://cloud.githubusercontent.com/assets/1645670/9277195/d45514ae-42a8-11e5-9ed4-7ee763bdd424.png)

Questions:
- I couldn't disable the "Any repo by URL" code, currently its hidden by the template. 
  - Is it required to do this in this MR? (probably it makes sense to refactor it in some other MR and to move it to the same structure like other importers)
  - If its required, can you please show/explain me the modules/logic/flow? (I was lost at some point in the flow between the modules)

Following points are implemented/todo:
- [x] Settings in the admin interface
- [x] DB migrations
- [x] Handling of import source configuration in "new project"
- [x] Spinach tests fixed/added

Something else?

artem-sidorenko committed:
Import sources: settings in the admin interface

artem-sidorenko committed:
Import sources: show only enabled sources

JobV commented:
@randx although I'm not a big fan of the feature, the implementation is nice. Problem is that we'd also have to adopt this for EE.

artem-sidorenko commented:
@JobV @randx let me try to make a deal:) if you would accept this one and the #9552, I would provide you both MRs for CE and EE.
So your efforts will be "only" to review the MRs and to maintain it in the future. What do you think?

artem-sidorenko commented:
- I took  a look to the spinach tests for "admin->settings", they are quite simple right now. Should I create/add/extend it to cover the import sources?
- Same for spec tests, should I create them?

randx commented:
@artem-sidorenko what happens if import for `GitHub` enabled but OAuth section is missing? Right now user see popup with link how to configure it.

artem-sidorenko commented:
@randx in the admin interface there is a help hint for admins with links to the help pages how to configure it (s. screenshot) . 
In the user interface the enabled sources with required but missing OAuth configuration are not displayed, as this leads to the user confusion (s. Issue 1): this steps require admin rights on the gitlab installation.

randx commented:
@artem-sidorenko ok. I am thinking about next case: people usually need import feature when they are new to GitLab. So I can easily imagine admin user does not find or understand `Import sources` option at first. Or admin can not know users would like to use it. At same time user does not see any import button (because OAuth is not configured) and does import manually even without knowing about all import features GitLab has.

artem-sidorenko commented:
@randx what about to display such oauth sources always, but to disable(and grey out?) the buttons in cases where its enabled but not configured? Or do you have some other idea?

randx commented:
@artem-sidorenko I think it should say something like `Ask administrator to enable it`. Otherwise its unclear why it is disabled

artem-sidorenko commented:
@randx so, we have following picture, please correct me if I'm wrong

user view (create new project):
- all import sources are enabled by default
- all enabled sources are displayed for the user
- sources with required oauth configuration display a modal window/view with a hint like "Ask administrator to enable it", without any details to do that

admin view:
- admin can enable/disable the import sources in the settings (like the current implementation above)
- admin can see the links to the help pages to the details how to configure oauth (like the current implementation above)

right?

randx commented:
> sources with required oauth configuration display a modal window/view with a hint like "Ask administrator to enable it", without any details to do that

@artem-sidorenko maybe better WITH details how to do that. If I am an admin I dont want to search for it

randx commented:
@artem-sidorenko or you already covered it with ?

> admin can see the links to the help pages to the details how to configure oauth

artem-sidorenko commented:
@randx ok, then I will keep similar like it [was](https://github.com/gitlabhq/gitlabhq/pull/9563/files#diff-306ad2ff78a4752e919d7f8000ba2d39L9), right?
Regarding the second question, yes "?" is pointing to the help page of gitlab installation (I did similar to the help page of in the visibility levels)

randx commented:
@artem-sidorenko cool

artem-sidorenko commented:
@randx may you please have a look? Admin interface is the same, changes are in the "new project" part+spinach tests


**New project action from user with admin rights:**
![new-project-admin](https://cloud.githubusercontent.com/assets/1645670/9357114/66a800b6-4685-11e5-9728-b9b2aac281e3.png)
(the same for Bitbucket and Gitlab.com)


**New project action from user without admin rights:**
![new-project-user](https://cloud.githubusercontent.com/assets/1645670/9357122/749bc2ca-4685-11e5-894b-ad2649ec043b.png)
(the same for Bitbucket and Gitlab.com)


**New project action with GitHub disabled:**
![new-project-github-disabled](https://cloud.githubusercontent.com/assets/1645670/9357134/7f758e06-4685-11e5-9c42-35ec24f61978.png)


Is this Ok?

artem-sidorenko committed:
Import sources: changelog

artem-sidorenko commented:
@randx I rebased it on the latest master and squashed the new changes to avoid "changes of changes". Please give me a ping if something needs to be done here or if I should squash it to one single commit

randx commented:
@artem-sidorenko looks good! I am going to merge this

artem-sidorenko commented:
@randx cool, thx:)

randx commented:
@artem-sidorenko welcome :) Thanks for good PR

artem-sidorenko commented:
@randx its possible because of good contribution guides and development kit from you guys :)

[We](https://github.com/Telekom-PD) have some further ideas/improvements for Gitlab/Gitlab CI and we will try to provide more PRs and hopefully reach at some point your PR quality wish "He loves ... a merge request that can be accepted without comments." :D

randx commented:
@artem-sidorenko awesome comment! :heart:
